staker/challenge-cache/cache.go

```diff
@@ -0,0 +1,242 @@
+// Copyright 2023-2024, Offchain Labs, Inc.
+// For license information, see https://github.com/offchainlabs/bold/blob/main/LICENSE
+/*
+* Package challengecache stores hashes required for making history commitments in Arbitrum BOLD.
+When a challenge begins, validators need to post Merkle commitments to a series of block hashes to
+narrow down their disagreement to a single block. Once a disagreement is reached, another BOLD challenge begins
+to narrow down within the execution of a block. This requires using the Arbitrator emulator to compute
+the intermediate hashes of executing the block as WASM opcodes. These hashes are expensive to compute, so we
+store them in a filesystem cache to avoid recomputing them and for hierarchical access.
+Each file contains a list of 32 byte hashes, concatenated together as bytes.
+Using this structure, we can namespace hashes by message number and by challenge level.
+
+Once a validator receives a full list of computed machine hashes for the first time from a validatio node,
+it will write the hashes to this filesystem hierarchy for fast access next time these hashes are needed.
+
+Example uses:
+- Obtain all the hashes for the execution of message num 70 to 71 for a given wavm module root.
+- Obtain all the hashes from step 100 to 101 at subchallenge level 1 for the execution of message num 70.
+
+	  wavm-module-root-0xab/
+		rollup-block-hash-0x12...-message-num-70/
+			hashes.bin
+			subchallenge-level-1-big-step-100/
+				hashes.bin
+
+We namespace top-level block challenges by wavm module root. Then, we can retrieve
+the hashes for any data within a challenge or associated subchallenge based on the hierarchy above.
+*/
+
+package challengecache
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var (
+	ErrNotFoundInCache    = errors.New("no found in challenge cache")
+	ErrFileAlreadyExists  = errors.New("file already exists")
+	ErrNoHashes           = errors.New("no hashes being written")
+	hashesFileName        = "hashes.bin"
+	wavmModuleRootPrefix  = "wavm-module-root"
+	rollupBlockHashPrefix = "rollup-block-hash"
+	messageNumberPrefix   = "message-num"
+	bigStepPrefix         = "big-step"
+	challengeLevelPrefix  = "subchallenge-level"
+)
+
+// HistoryCommitmentCacher can retrieve history commitment hashes given lookup keys.
+type HistoryCommitmentCacher interface {
+	Get(lookup *Key, numToRead uint64) ([]common.Hash, error)
+	Put(lookup *Key, hashes []common.Hash) error
+}
+
+// Cache for history commitments on disk.
+type Cache struct {
+	baseDir string
+}
+
+// New cache from a base directory path.
+func New(baseDir string) (*Cache, error) {
```

Comment:
I like to separate New and Init. New creates the struct, while Initialize can get a context for parameter and it may perform longer actions, like in this instane - calling os and checking things.

It's not a must, but helps organiztion.